project/src/test/java/com/google/sps/servlets/FeedbackServletTest.java

```diff
@@ -0,0 +1,131 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.argThat;
+import static org.mockito.Mockito.atLeastOnce;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import com.google.common.collect.ImmutableList;
+import com.google.sps.data.DataAccessor;
+import com.google.sps.data.UserFeedback;
+import com.google.sps.data.UserVerifier;
+import java.util.Optional;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class FeedbackServletTest {
+
+  private static final HttpServletRequest REQUEST = mock(HttpServletRequest.class);
+  private static final HttpServletResponse RESPONSE = mock(HttpServletResponse.class);
+  private static final String ID_TOKEN = "abcde";
+  private static final String USER_ID = "12345";
+  private static UserVerifier mockUserVerifier;
+  private static DataAccessor mockDataAccessor;
+  private FeedbackServlet servlet;
+
+  @Before
+  public void setUp() throws Exception {
+    mockUserVerifier = mock(UserVerifier.class);
+    mockDataAccessor = mock(DataAccessor.class);
+    servlet = new FeedbackServlet();
+    servlet.init(mockUserVerifier, mockDataAccessor);
+  }
+
+  @Test
+  public void doPost_validToken_updatesFeedback() throws Exception {
+    mockValidRequestParams(ID_TOKEN,
+        "place1,place2,place3" /* recommendedPlaces */,
+        "place1" /* chosenPlace */,
+        "false" /* tryAgain */);
+    when(mockUserVerifier.getUserIdByToken(ID_TOKEN)).thenReturn(Optional.of(USER_ID));
+
+    servlet.doPost(REQUEST, RESPONSE);
+
+    // The parameters here must match those that are mocked in the request.
+    // We can't compare instances of UserFeedback, because of the "feedbackTimeInMillis" attribute.
+    verify(mockDataAccessor).updateUserFeedback(argThat(feedback -> {
```

Comment:
FYI: you can always create a custom ArugmentMatcher!
Then you can simplify the syntax to:
`updateUserFeedback(argThat(MatchesUserFeedback(USER_ID, ...)))`

But since you only test this once, it's an overkill here :)

Reply:
Right, I used what Tal did as an example, and I saw that's how she did it. I figured that since it's used once, it would we simpler this way